volatility3/framework/plugins/windows/vadtree.py

```diff
@@ -0,0 +1,143 @@
+                                         element_type = int,
+                                         description = "Process IDs to include (all other processes are excluded)",
+                                         optional = True)
+        ]
+
+    @classmethod
+    def get_heaps(cls, proc) -> List[int]:
+        """
+        """
+        try:
+            return [proc.get_peb().ProcessHeaps.dereference()]
+        except exceptions.InvalidAddressException:
+            #vollog.log()
+            return renderers.UnreadableValue()
+
+    @classmethod
+    def get_modules(cls, proc) -> List[int]:
+        """
+        """
+        modules = []
+        for mod in proc.load_order_modules():
+            try:
+                modules.append(mod.DllBase)
+            except exceptions.InvalidAddressException:
+                #vollog.log()
+                continue
+        return modules
+
+    def get_stacks(self, proc: interfaces.objects.ObjectInterface) -> List[int]:
+        #TODO Exception Processing
+        stacks = []
+
+        kernel = self.context.modules[self.config['kernel']]
+        layer_name = kernel.layer_name
+
+        kvo = self.context.layers[layer_name].config['kernel_virtual_offset']
+        ntkrnlmp = self.context.module(kernel.symbol_table_name, layer_name=kernel.layer_name, offset=kvo)
+        tleoffset = ntkrnlmp.get_type("_ETHREAD").relative_child_offset("ThreadListEntry")
+
+        ethread = ntkrnlmp.object(object_type="_ETHREAD",
+                                    offset=proc.ThreadListHead.Flink - tleoffset,
+                                    absolute=True)
+
+        while(True):
+            stacks.append(ethread.Tcb.StackBase)
+            ethread = ntkrnlmp.object(object_type="_ETHREAD",
+                                        offset=ethread.ThreadListEntry.Flink - tleoffset,
+                                        absolute=True)
+            
+            if(ethread.ThreadListEntry.Flink == proc.ThreadListHead.Blink):
+                break
+
+        return stacks
+    
+    def get_type(self, proc, vad) -> str:
+        heaps = self.get_heaps(proc)
+        modules = self.get_modules(proc)
+        stacks = self.get_stacks(proc)
+        
+        type = renderers.NotApplicableValue()
+        
+        if(vad):
+            if vad.get_start() in heaps:
+                type = "Heap"
+            elif vad.get_start() in modules:
+                type = "Module"
+            elif vad.get_start() in stacks:
+                type = "Stack"
+            else:
+                try:
+                    if vad.FileObject.FileName:
+                        type = "File"
+                except AttributeError:
+                    pass
+        
+        return type
+
+    def _generator(self, procs) -> Iterator[Tuple]: 
+        for proc in procs:            
+            levels = {}
+
+            for vad in vadinfo.VadInfo.list_vads(proc):
+                level = levels.get(vad.get_parent(), -1) + 1
```

Comment:
So, this line and the next should probably be using `& self.context.layers[vad.vol.layer_name].address_mask` to ensure consistency.  This can also be used for the output in vadwalk to produce similar looking results.

Reply:
From what I understand, should I think that implementing `VADWalk` and `VADTree`, which are consistent through `address_mask`, is a good direction? :)
Then, I think we do not calculate the offset of VAD strangely.

Reply:
After testing it, I think it's working well! Thank you for your help.
If there is no problem, I will close #731